src/radical/pilot/utils/serializer.py

```diff
@@ -0,0 +1,165 @@
+
+'''
+The serializer should be able to (de)serialize information that we want
+to send over the wire from the client side to the agent side via
+1- ZMQ
+2- MongoDB
+
+we except:
+    1- Callables with and without dependecies.
+    2- Non-callables like classes and other python objects
+'''
+__copyright__ = 'Copyright 2022, The RADICAL-Cybertools Team'
+__license__   = 'MIT'
+
+import os
+import dill
+import pickle
+import codecs
+import tempfile
+
+
+_obj_dir       = tempfile.gettempdir()
+_obj_file_name = 'rp_obj.pkl'
+_obj_file_path = os.path.join(_obj_dir, _obj_file_name)
+
+
+# ------------------------------------------------------------------------------
+#
+def serialize_obj(obj):
+    '''
+    serialize object
+    '''
+    result    = None
+    exception = None
+
+    if callable(obj):
+        try:
+            result = dill.dumps(obj)
+        except Exception as e:
+            exception = e
+
+        # if we fail, then pikle it by reference
+        if result is None:
+            # see issue: https://github.com/uqfoundation/dill/issues/128
+            try:
+                result = dill.dumps(obj, byref = True)
+            except Exception as e:
+                exception = e
+
+    else:
+        try:
+            result = dill.dumps(obj, recurse = True)
+        except Exception as e:
+            exception = e
+
+        if result is None:
+            # see issue: https://github.com/uqfoundation/dill/issues/128
+            try:
+                result = dill.dumps(obj, byref = True)
+            except Exception as e:
+                exception = e
+
+    if result is None:
+        raise Exception("object %s is not serializable") from exception
+
+    return  result
+
+
+# ------------------------------------------------------------------------------
+#
+def serialize_file(obj):
```

Comment:
What is this method actually used for?  I have not seen it called anywhere in the code.

Reply:
It's a utility that might/might not be used. The optimal approach (for this purpose I implemented this) is that if we fail to serialize to byte object in memory then we switch to serialize an object to a file. This feature is not implemented yet :(.